Add support for date-based restart frequency

jo-basevi opened the pull request:
Add support for date-based restart pruning in Payu - should close #358.

As suggested by @aidanheerdegen [here](https://github.com/payu-org/payu/issues/358#issue-1856663402), an option is to support a subset of pandas date offset frequency aliases:
| Unit        | Description | 
| :---        |    :----   | 
| MS          | Month start frequency 
| YS          | Year start frequency 
| M           | Month frequency
| Y           | Year frequency 
| W           | Weekly frequency
| D           | Day frequency
| H           | Hour frequency
| T           | Minute frequency
| S           | Second freqeuncy  

So for example, if in `config.yaml`, 
```
restart_freq: 5YS
```
Then the earliest restart of the year, every 5 years from the first restart will be retained.

During archive, the function `Experiment.prune_restarts()` would inspect each restart directory and return a list of restart directories that can removed using integer or date-based restart frequency. For date-based frequency, individual restart directory paths are passed to the model specific driver that can parse the restart files and return a cftime (or standard) datetime object.

In this branch, `access-om2` has the `model.get_restart_datetime()` function implemented, which calls the mom's `get_restart_datetime()` which parses the `ocean_solo.res` file for the final datetime of the restart. I used this file as it is what is used in COSIMA's [tidy_restarts script](https://github.com/COSIMA/1deg_jra55_ryf/blob/master/tidy_restarts.py) and it also contains information on what calendar it is using. 

Currently the month (M) and year (Y) frequencies are defined similar to adding relativedelta's month and years. So for example if a datetime was 15/01/2000, after adding an offset of '6M' it would be 15/07/2000. I have only just noticed that 'M' in pandas documentation is actually the end of the month.. Would the end of the month/year be more useful? Also, with the different cftime calendars M and Y could give unexpected results i.e with different day months. For example, I've got some logic for cftime's`360_day`, `noleap` and `all_leap` calendars but nothing yet for the `julian` calendar. Also would the start of month/year (MS/YS) frequencies be sufficient- Is M and Y frequencies even needed?

Week, day, hour, minute frequencies may not be super useful or even necessary but they were easy to add as adding timedeltas is supported with both cftime and standard datetime calendars. 

The first restart datetime is used as a point of reference for adding the first frequency interval.  Then the next restart with a datetime at or after this checkpoint is kept and then becomes the reference when adding the next time interval. I think using the most recent "kept" datetime as a reference for the next checkpoint could be better than strict regular intervals from the earliest datetime. As the first restart could eventually be lost due to scratch timeouts and using "YS"/"MS" would still keep the earliest restarts in each year/month. 

As one of the motivations for date-based restart frequencies was to make syncing restarts to a remote archive easier as payu would know what restarts could be stored permanently. A way to check if a restart would eventually be removed would be to see if the restart directory was in the list returned by `prune_restarts(from_n_restart=restart_history - 1, to_n_restart:self.counter)` where `restart_history` is the config.yaml or default value of how many latest restarts are kept.

Any feedback, questions or suggestions is much appreciated!

aidanheerdegen reviewed (changes requested):
This is great! Thanks for the excellent work.

Sorry it has taken so long to finish this review. 

I've still not tried it out with a model, which I would like to do, but I didn't want to hold you up and figured this review was long enough as it was.

jo-basevi commented on payu/calendar.py:
```diff
@@ -86,3 +89,109 @@ def calculate_leapdays(init_date, final_date):
+            'T': add_timedelta_fn(datetime.timedelta(minutes=1)),
+            'S': add_timedelta_fn(datetime.timedelta(seconds=1))
+        }
+        assert unit in supported_datetime_offsets, f"payu: error: unsupported datetime offset: {unit}"
```
Thanks, that is good to know!

jo-basevi commented on payu/experiment.py:
```diff
@@ -995,6 +980,56 @@ def sweep(self, hard_sweep=False):
+    
+    def prune_restarts(self, from_n_restart=0, to_n_restart=None):
+        """Returns a list of restart directories that can be pruned"""
+        restart_freq = self.config.get('restart_freq', 5)
```
Oops that was my error, thanks for picking that up!

pep8speaks commented:
Hello @jo-basevi! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`payu/calendar.py`](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py):

> [Line 98:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L98): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 104:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L104): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 106:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L106): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 107:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L107): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 109:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L109): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 113:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L113): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 115:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L115): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 116:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L116): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 120:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L120): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 121:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L121): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 128:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L128): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 133:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L133): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 134:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L134): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 79 characters)
> [Line 136:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L136): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)
> [Line 137:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L137): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 139:37](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L139): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 140:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L140): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)
> [Line 141:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L141): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 143:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L143): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 146:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L146): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 152:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L152): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 154:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L154): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 79 characters)
> [Line 155:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L155): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 79 characters)
> [Line 167:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L167): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 173:5](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L173): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 175:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L175): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 177:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L177): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 180:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L180): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 181:41](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L181): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 181:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L181): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 183:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L183): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 79 characters)
> [Line 184:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L184): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 185:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L185): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 79 characters)
> [Line 186:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L186): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 79 characters)
> [Line 188:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L188): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 192:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L192): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 194:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L194): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 197:55](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/calendar.py#L197): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`payu/experiment.py`](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py):

> [Line 750:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L750): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 983:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L983): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 987:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L987): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 991:33](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L991): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 995:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L995): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 79 characters)
> [Line 997:46](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L997): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 999:57](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L999): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1003:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1003): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 1004:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1004): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 1006:56](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1006): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1020:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1020): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1022:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1022): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 1022:85](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1022): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1023:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1023): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1024:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1024): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 1027:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1027): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (136 > 79 characters)
> [Line 1028:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1028): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 1030:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1030): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 1031:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/experiment.py#L1031): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`payu/models/accessom2.py`](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/accessom2.py):

> [Line 95:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/accessom2.py#L95): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 96:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/accessom2.py#L96): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 79 characters)

* In the file [`payu/models/model.py`](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/model.py):

> [Line 461:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/model.py#L461): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 463:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/model.py#L463): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 79 characters)
> [Line 463:120](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/model.py#L463): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 465:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/model.py#L465): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`payu/models/mom.py`](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py):

> [Line 13:14](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L13): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 231:5](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L231): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 232:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L232): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 79 characters)
> [Line 236:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L236): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 245:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L245): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 247:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L247): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 248:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L248): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)
> [Line 248:105](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L248): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 249:33](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L249): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 250:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L250): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 250:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/payu/models/mom.py#L250): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



* In the file [`test/test_calendar.py`](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py):

> [Line 8:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L8): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 11:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (95 > 79 characters)
> [Line 16:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L16): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 22:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L22): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 26:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L26): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 27:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L27): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 28:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L28): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 29:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L29): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 30:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L30): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 32:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L32): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 34:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L34): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 42:33](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L42): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 43:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L43): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 47:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L47): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 48:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L48): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 49:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L49): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 50:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L50): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 51:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L51): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 53:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L53): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 55:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L55): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 63:32](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L63): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 64:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L64): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 68:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L68): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 69:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L69): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 71:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L71): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 78:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L78): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 79:53](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L79): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 81:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L81): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 82:34](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L82): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 83:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L83): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 87:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L87): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 89:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L89): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 99:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L99): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 101:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L101): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 106:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L106): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 107:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L107): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 112:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L112): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 116:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L116): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 117:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L117): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 79 characters)
> [Line 119:1](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L119): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 122:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L122): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)
> [Line 125:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L125): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 126:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L126): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 127:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L127): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 128:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L128): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 79 characters)
> [Line 129:80](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L129): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 132:50](https://github.com/payu-org/payu/blob/3033d174575ab8126ade9c47155fca1b33e55214/test/test_calendar.py#L132): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

jo-basevi commented on test/test_calendar.py:
```diff
@@ -0,0 +1,132 @@
+                               calendar='noleap'),
+        '4M': cftime.datetime(year=2001, month=2, day=28, hour=10, minute=5, second=2,
+                              calendar='noleap')
+    }
```
Thanks for the notes! I ended up rewriting the tests as the parameterisation simplified them a lot!

jo-basevi commented on payu/calendar.py:
```diff
@@ -86,3 +89,109 @@ def calculate_leapdays(init_date, final_date):
+    
+    if isinstance(initial_dt, cftime.datetime):
+        year = initial_dt.year + ((initial_dt.month + n - 1) // 12)
+        month = (initial_dt.month + n - 1) % 12 + 1
```
Thanks, I rewrote it so the `year_to_add` variable was used to calculate how the `months_to_add`.

jo-basevi commented on payu/calendar.py:
```diff
@@ -86,3 +89,109 @@ def calculate_leapdays(init_date, final_date):
+        
+        if initial_dt.calendar == "all_leap":
+            max_day_in_month[2] = 29 # Every year is a leap year
+            day = initial_dt.day if initial_dt.day <= max_day_in_month[month] else max_day_in_month[month]
```
Ok, thanks for all those comments! I've removed that logic for the `M` and `Y` offsets that only supported `360_day`,  `noleap`/`365_day`, 'all_leap'/`366_day`.  (Note when creating cftime's with `365_day` calendar, accessing the calendar attribute gives 'noleap' - Similar for 'all_leap'/`366_day`)

'MS', 'YS' hopefully will be sufficient for month and year offsets. As they are getting truncated to start of month/year, it supports all cftime calendars.

I've changed the `gregorian` to `proleptic_gregorian` calendar in mom's function for parsing restart date so it works for dates before 1582.

aidanheerdegen replied:
> I've changed the `gregorian` to `proleptic_gregorian` calendar in mom's function for parsing restart date so it works for dates before 1582.

Just to be clear, MOM5 (or to be specific, the version of FMS in MOM5) doesn't actually properly support `gregorian` (or `standard` to use it's ... standard ... name). But newer versions of FMS do calculate it correctly. As this is only an `accessom2` driver it is fine to say what it does, not what it purports to do I guess, but I'm still a bit nervous! :)

jo-basevi commented on payu/models/mom.py:
```diff
@@ -224,3 +226,25 @@ def create_mask_table(self, input_nml):
+        year, month, day, hour, minute, second = [int(i) for i in last_date_line[:6]]
+        return cftime.datetime(year=year, month=month, day=day, hour=hour, minute=minute, second=second, 
+                                calendar=calendar)
+        
```
I've added tests that generate `ocean_solo.res` files, to test mom's parsing of restart directories (in `test_mom.py`). I reused functions to set up restart directories and files in `test_prune_restarts.py` to have wider integration tests for `Experiment.prune_restarts()`. In the `test_prune_restarts.py`, I've also used the `access-om2`'s model just to also check that is working as expected. 

The alternative in `test_prune_restarts.py` is to mock out the Model class which I could also figure out how to do that in python if you think it's better.

jo-basevi commented:
Main changes were removing `M` and `Y` offset options which simplified the code quite a bit 
as all cftime calendars are supported with `YS` - Year start and `MS` - Month start (i.e don't have to worry about the end of month logic..).

Other changes were rewriting the calendar tests and added integration-like tests for parsing MOM restart files and the `prune_restarts()` method in the Experiment class.

**Other Notes**
I removed this from the documentation:   
    " Intermediate restarts are not deleted until a permanently archived restart
   has been produced. For example, if we have just completed run ``11``, then
   we keep ``restart004``, ``restart009``, ``restart010``, and ``restart011``.
   Restarts 10 through 13 are not deleted until ``restart014`` has been saved."

   Because the logic previously was to not delete the last `restart_history` number of restarts. Intermediate restarts were not deleted until a permanently saved restart was only automatically the case when `restart_freq` was 5 or less as the default `restart_history` was also 5. Or if `restart_history` was also configured to be greater than or equal to `restart_freq`. I could change it so `restart_history` is set to be minimum of `restart_freq` to guarantee immediate restarts aren't pruned too early for integer restart frequency and add it back into the documentation?

Other things I got caught out on was the difference between `12MS` and `1YS`, as the logic uses the runtime of earliest restart as a reference (not the model start time). Say for example, there were runs with runtime of 1 month, with a mom model start time of 1900/01/01:
(run_date, restart_directory)
(1900/02/01, restart000)
(1900/03/01, restart001)
(1900/04/01, restart002)
...

If using frequency `12MS`, the restarts that would be saved are:
(1900/02/01, restart000)
(1901/02/01, restart012)
(1901/02/01, restart024)
...

While `1YS` would keep 
(1900/02/01, restart000)
(1901/01/01, restart011)
(1901/01/01, restart023)
...

The other thing to keep in mind is that the code currently uses the next kept restart as a reference for the next time interval. This means that restarts saved may not be uniformly a fixed frequency from the first restart if there's irregular time steps. 

For `restart_freq: 6MS`, theres a minimum 6 months between each saved restart. So with using the restart's list from above, the restarts saved would be:
(1900/02/01, restart000)
(1900/09/01, restart002) (earliest restart after 6 months)
(1901/03/01, restart004) (6 months after previous kept restart)

If instead there was a fixed frequency from the first restart, the restarts kept would be:
(1900/02/01, restart000)
(1900/09/01, restart002)
(1901/02/01, restart003) 
If restart000 was lost somehow then the logic would break as restart002 would be the new reference and restart003 would be purged in a subsequent run through..

aidanheerdegen commented:
> **Other Notes** I removed this from the documentation: " Intermediate restarts are not deleted until a permanently archived restart has been produced. For example, if we have just completed run `11`, then we keep `restart004`, `restart009`, `restart010`, and `restart011`. Restarts 10 through 13 are not deleted until `restart014` has been saved."
> 
> Because the logic previously was to not delete the last `restart_history` number of restarts. Intermediate restarts were not deleted until a permanently saved restart was only automatically the case when `restart_freq` was 5 or less as the default `restart_history` was also 5. Or if `restart_history` was also configured to be greater than or equal to `restart_freq`. I could change it so `restart_history` is set to be minimum of `restart_freq` to guarantee immediate restarts aren't pruned too early for integer restart frequency and add it back into the documentation?

Good catch. Yes i think `restart_history` used should be `min(restart_history, restart_freq)`. It doesn't make sense to keep five previous restarts if the `restart_freq` is 2. So that would mean reinstating that bit in the docs, though it does need editing to ay it uses `restart_history` and keeps `min(restart_history, restart_freq)` of the most recent history files.

How does this affect date based pruning though? We can keep `restart_history` as an integer (just keep the `n` most recent restarts`) but then what does `min(restart_history, restart_freq)` mean? Would you translate `restart_freq` into an integer number of runs? Not sure how they could be determined.

> Other things I got caught out on was the difference between `12MS` and `1YS`

So this is just something that needs to be kept in mind by the user?

> The other thing to keep in mind is that the code currently uses the next kept restart as a reference for the next time interval. This means that restarts saved may not be uniformly a fixed frequency from the first restart if there's irregular time steps.
> 
> For `restart_freq: 6MS`, theres a minimum 6 months between each saved restart. So with using the restart's list from above, the restarts saved would be: (1900/02/01, restart000) (1900/09/01, restart002) (earliest restart after 6 months) (1901/03/01, restart004) (6 months after previous kept restart)

The restarts from above were monthly, so I'm confused by `(1900/02/01, restart000) (1900/09/01, restart002)` shouldn't it be (1900/02/01, restart000) (1900/09/01, restart007)?

> If instead there was a fixed frequency from the first restart, the restarts kept would be: (1900/02/01, restart000) (1900/09/01, restart002) (1901/02/01, restart003) 

Which is the same as above .. no?

> If restart000 was lost somehow then the logic would break as restart002 would be the new reference and restart003 would be purged in a subsequent run through..

I'm still not sure those restart numbers are correct, or if they are I understand why, so it's a bit hard to make an informed comment on this bit.

Just some general questions:

1. If a user would like to always save restarts at the beginning of a year (or as close as possible) is that possible?
2. Do we need to save some state about previous inferred restart frequencies to be able to warn users if they make a change that could be deleterious?

jo-basevi commented:
>I'm still not sure those restart numbers are correct, or if they are I understand why, so it's a bit hard to make an informed comment on this bit.

Sorry that's my bad!! I deleted the list of restarts I was talking about. So here is the list of restarts I was referring to:
```
(1900/02/01, restart000)
(1900/07/01, restart001)
(1900/09/01, restart002)
(1901/02/01, restart003)
(1901/03/01, restart004)
...
```

aidanheerdegen commented on payu/experiment.py:
```diff
@@ -995,6 +980,56 @@ def sweep(self, hard_sweep=False):
+        restarts_to_prune = []
+        if self.repeat_run:
+            # TODO: Previous logic was to prune all restarts if self.repeat_run - is that still the case?
+            restarts_to_prune = [os.path.join(self.archive_path, restart) for restart in restarts]
```
I was going to use MOM6 for this, but as you pointed out this is not supported by the MOM6 driver. Having said which, they both use FMS, so MOM6 has an `ocean_solo.res` file.

So ... if this was an FMS driver it would work with any model that sub-classes `Fms`, which includes both `mom5` and `mom6`.

In which case I might rescind my comment about `proleptic_gregorian`.

jo-basevi replied:
Ok, so move the `mom`'s `get_restart_datetime()` to the fms model driver? The other model that subclasses of `Fms` is GOLD, so would it be worth checking if there is an `ocean_solo.res` exists and if not raise an not implemented error?

jo-basevi commented:
> If a user would like to always save restarts at the beginning of a year (or as close as possible) is that possible?

Yes, using 1YS would save the restart where current model run time is the earliest of each year. 

> Do we need to save some state about previous inferred restart frequencies to be able to warn users if they make a change that could be deleterious?

Would that require saving the previous restart frequency as an enviroment variable? Changing to a longer
date-based period or a shorter integer frequency would not result in deletions. Though it would be tricky comparing an integer frequency with a datebased frequency..

As archive() runs after each experiment run, if there is greater than 1 restart to prune, that will indicate that either:
- restart_frequency is set to a shorter date-based or  higher integer value
- or that restart_history is changed to a lower value.

Instead of saving previous state, during setup and if archive is enabled, I could inspect what restarts it would prune (but not actually delete them..). If `restart_freq` and `restart_history` are the same, it should be an empty list at that stage. If not, emit a warning that the config has changed, and more restarts will be deleted at the next time archive is run. 

How does that sound? Running that check in setup would also have the benefit for any warnings regarding invalid `restart_freq` values.

jo-basevi commented:
>Good catch. Yes i think restart_history used should be min(restart_history, restart_freq). It doesn't make sense to keep five previous restarts if the restart_freq is 2. So that would mean reinstating that bit in the docs, though it does need editing to ay it uses restart_history and keeps min(restart_history, restart_freq) of the most recent history files.
>How does this affect date based pruning though? We can keep restart_history as an integer (just keep the n most recent restarts) but then what does min(restart_history, restart_freq)mean? Would you translaterestart_freq` into an integer number of runs? Not sure how they could be determined.

Ok, an idea for an simple algorithm pattern to keep intermediate restarts is:
```python
restarts_to_prune = []
intermediate_restarts = []
for restart in restarts:
    if restart_should_be_pruned():
        # keep track of intermediate restarts 
        intermediate_restarts.append(intermediate_restarts)
    
    else:  # restart should be saved
        # add prior intermediate restarts to the list of restarts_to_prune
        restarts_to_prune.extend(intermediate_restarts)
        intermediate_restarts = []
```

This will work for both integer and date_based frequency and wouldn't require a `restart_history` config variable at all.. 
If `restart_history` is useful to keep, it could act as an override option to ensure that the last `n` restarts are always kept?

aidanheerdegen commented:
> How does that sound? Running that check in setup would also have the benefit for any warnings regarding invalid `restart_freq` values.

That sounds like a great idea, thanks. I definitely think there is potential for Bad Things<sup>TM</sup> to happen, and checks and emitting information to the user is a good idea. 

I definitely can see a use-case for someone to decide to change their `restart_frequency` when their experiment progresses. This could be to reduce disk usage or just because they realise they don't need so many. 

But I think we want to make this sort of mass-deletion an interactive only option. So warn that lots of restarts would be the result of the next `archive`, so make them use a `-f/--force` flag to make it happen, otherwise no restarts are pruned? 

Specifying flags isn't something you can do as part of an automated run, so by default it has to be done interactively.

jo-basevi commented:
*New Changes*

**Support for both mom5 and mom6 models**
- Move mom driver restart parsing method to fms driver so it can work with mom6 and mom models (as both have the ocean_solo.res file). The other fms model subclass is GOLD. I’m not too sure whether it has similar file or not.. So the method checks first that the file exists and raises a descriptive error if not. 

**Intermediate restarts and restart_hisory**
- Intermediate restarts can be kept with both integer and date-based restart pruning 
- However, if `restart_history` is defined, it works as it did previously and only keeps `restart_history` number of restarts. This would be useful for only keeping more or less restarts than intermediate restarts. 

**Force restart pruning**
- At the end of setup, a scan of what restarts would be pruned is run. This will error out if there's any invalid configuration (i.e. unsupported values, or if restart files can't be parsed). It will also warn if more restarts than expected would be pruned, and whether `--force-prune-restarts` needs to be added to next run/archive. This would be the case if `restart_history` or `restart_freq` has been changed to values which would prune more restarts, or if auto archiving is suddenly enabled.
- Added `--force-prune-restarts` flag to allow greater deletions of restarts, otherwise no restarts will be deleted. This can be run as `payu run --force-prune-restarts` or `payu archive --force-prune-restarts` - if archive is being run manually.

aidanheerdegen commented on test/test_prune_restarts.py:
```diff
@@ -0,0 +1,223 @@
+        print(e)
+
+
+@pytest.fixture(autouse=True)
+def teardown():
+    # Run test
+    yield
+
+    # Remove any created restart files
+    remove_expt_archive_dirs(type='restart')
```
Did not know about this. Very cool.

jo-basevi replied:
Yeah, it was useful in this case!

aidanheerdegen commented on test/test_prune_restarts.py:
```diff
@@ -0,0 +1,223 @@
+         [1, 2, 3, 4, 5, 6, 7, 8, 9, 10,
+          12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]),
+        (1, 1, []),
+        (5, 5, [1, 2, 3, 4, 6, 7, 8, 9, 11, 12, 13, 14, 16, 17, 18, 19]),
+        (5, None, [1, 2, 3, 4, 6, 7, 8, 9, 11, 12, 13, 14, 16, 17, 18, 19])
```
I guess I misunderstood, with `restart_history: None` I'd expect it to prune `21` as well. So by default all intermediate files are retained until a restart is pruned? This is good behaviour by default, so I'm not suggesting changing it, just wanted to confirm this.

In any case it would be prudent, and pretty easy, to include a case where `restart_history` is < `restart_freq` and vice_versa for the integer pruning options. Maybe make one that is a weird number, like 3 or 7?

jo-basevi replied:
Yes, the default behaviour if there's no `restart_history` is to keep intermediate restarts. So in this case the permanently archived restarts are 0, 5, 10, 15, 20, and intermediate restarts are 21, 22, 23, 24. The intermediate restarts will be deleted once there's restart 25. 

Thanks for the suggestion about the more tests- I've added in those cases for `restart_history` being 3 and 7

aidanheerdegen reviewed (approved):
Looks great, thanks for all the work. 

A couple of comments, maybe a change, but otherwise good to go I think.

aidanheerdegen commented:
> Move mom driver restart parsing method to fms driver so it can work with mom6 and mom models (as both have the ocean_solo.res file). The other fms model subclass is GOLD. I’m not too sure whether it has similar file or not.. So the method checks first that the file exists and raises a descriptive error if not.

I was wrong when I said this was FMS based, I mean, they're all FMS based, and so end up using the same convention, but the naming of the file is in the model driver code, not FMS. Regardless I think it is the right decision to put the code in the FMS driver for the moment. 

I guess longer term the best option might be to put it in a mixin class and inherit that from the models that use `ocean_solo.res`.

aidanheerdegen commented on docs/source/config.rst:
```diff
@@ -189,16 +189,29 @@ configuration.
+   able to parse restarts files for a datetime.
+
+``restart_history``
+    Specifies the number of latest restart files to save
```
Damn! Forgot to say that we need to flag that date based restart pruning is not available for all models.

jo-basevi replied:
Thanks for the suggestion! I had added a note on the lines above that date-based `restart_freq` is currently only supported for ACCESS-OM2, MOM6, and MOM5. Would that be sufficient?

aidanheerdegen replied:
Yep, looks good. Thanks.

jo-basevi commented on payu/experiment.py:
```diff
@@ -997,6 +995,127 @@ def sweep(self, hard_sweep=False):
+                    previous_intermediate_restarts = intermediate_restarts
+                    intermediate_restarts = []
+
+        if ignore_intermediate_restarts:
+            # Return all restarts that'll eventually be pruned
+            restarts_to_prune.extend(intermediate_restarts)
+            return restarts_to_prune
```
Yes that is currently the case, I was planning on using that flag in the future for the remote archive rsync method to check what restarts would be permanently archived/ or eventually pruned.

jo-basevi commented on payu/experiment.py:
```diff
@@ -997,6 +995,127 @@ def sweep(self, hard_sweep=False):
+
+        # List all and sort restart directories in archive
+        restarts = self.list_output_dirs(output_type='restart')
+        restarts.sort(key=lambda d: int(d.lstrip('restart')))
```
Ok good idea, I've added the sort to the `list_output_dirs` method. I've also tweaked the regex to allow counters > 999

jo-basevi commented:
Thanks for the review @aidanheerdegen! I've added in the suggested changes and the small fix to counter > 999. Should close #372

aidanheerdegen reviewed (approved):
Looks great!

aidanheerdegen commented:
I've replied where appropriate and resolved all the conversations, so good to merge.
